### ISSUE TITLE
Logical focus- and reading order between markers and popups

### ISSUE THREAD

**Issue by danallsop** [ScottLogic/jumpstart-Leaflet#345]
> **Migrated from:** https://github.com/Leaflet/Leaflet/issues/8113
> **Original author:** danallsop
> **Original created:** 2022-04-04T14:54:18Z
---

_Raising this issue on behalf of danallsop._

**Is your feature request related to a problem? Please describe.**

The focus order ("tab order") between markers (or other elements bound to popups) and popups is not logical.

This is due to markers and popups existing in different subtrees of the DOM (`leaflet-marker-pane` and `leaflet-popup-pane`).

A [logical focus order is desirable](https://www.w3.org/WAI/WCAG21/Understanding/focus-order.html) so that (all) users can:

> encounter information in an order that is consistent with the meaning of the content and can be operated from the keyboard. This reduces confusion by letting users form a consistent mental model of the content.

**Describe the solution you'd like**

danallsop's https://github.com/Leaflet/Leaflet/issues/7968#issuecomment-1069147811:

> - Ensure that the popup appears immediately below the marker in the DOM

**Additional context**

- Current behavior fails WCAG [Success Criterion 2.4.3 Focus Order](https://www.w3.org/TR/WCAG21/#focus-order)
- [web.dev: DOM order matters](https://web.dev/dom-order-matters/)

**Comment by danallsop** [comment]
## Comments (3 total)

---
> **Comment by:** danallsop **on** 2022-04-04T14:56:44Z

A proper DOM order is very important, however I believe it would require a big architectural change (basically asking to remove `leaflet-marker-pane`?), which I'm not sure is realistic. At least not in the short term.

Otherwise, if this request (of moving popups after markers in the DOM) is unlikely to happen:

- https://github.com/Leaflet/Leaflet/issues/8115 would correct the _focus_ order.
  (but only when opening/closing popups, not for <kbd>tab</kbd> navigation in general)
- [`aria-flowto`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-flowto) could help correct the _reading_ order.
---
> **Comment by:** danallsop **on** 2023-09-11T14:51:12Z

Yes, this is a major problem for keyboard and screen-reader accessibility. I've been getting around this with a `'popupopen'` event-handler on the Marker, where I simply _move_ the Popup-container's DOM element so that it's the next-sibling of the Marker-icon's DOM element (like `event.sourceTarget._icon.insertAdjacentElement('afterend', event.popup._container)`). Also have to give the Popup-container element a very high z-index (to beat out all the Markers). This gets the natural tab-order right, and for my purposes it seems to work just fine, though I'd imagine that for some purposes this might result in broken functionality.
---
> **Comment by:** danallsop **on** 2025-10-30T22:05:48Z

I just ran into this same issue while setting up a map and doing accessibility testing.  For me it works fine if there's only one pin on the map, but as soon as you have multiple pins the tab order gets messed up.  Especially if you have more than a couple pins, it's so messed up as to seem completely useless, at first I didn't realize it was just the tab order and thought certain elements were completely inaccessible via tabbing.  The last activity I see here seems to have marked this as "not prioritized."  Is there any sort of plan for when this might be addressed?
